### PR TITLE
refactor: Migrate store to TypeScript

### PR DESCRIPTION
**Pull request by Gudahtt**
**Development & PR Process**
1. Follow MetaMask [Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/coding_guidelines/CODING_GUIDELINES.md)
2. Add `release-xx` label to identify the PR slated for a upcoming release (will be used in release discussion)
3. Add `needs-dev-review` label when work is completed
4. Add the appropiate QA label when dev review is completed
    - `needs-qa`: PR requires manual QA.
    - `No QA/E2E only`: PR does not require any manual QA effort. Prior to merging, ensure that you have successful end-to-end test runs in Bitrise. 
    - `Spot check on release build`: PR does not require feature QA but needs non-automated verification. In the description section, provide test scenarios. Add screenshots, and or recordings of what was tested.
5. Add `QA Passed` label when QA has signed off (Only required if the PR was labeled with `needs-qa`)
6. Add your team's label, i.e. label starting with `team-` (or `external-contributor` label if your not a MetaMask employee)

**Description**

The Redux store module and the root reducer have been migrated to TypeScript. We now have the start of a root state object for the Redux state. Most of the reducers still aren't typed, or there are type errors, but it's a start. The missing pieces are typed as `any` for now.

The TypeScript conversion introduced some lint errors caused by the pattern of using destructuring to omit object properties. The rule has been updated to allow this pattern (it's already allowed on the latest shared ESLint config).

The store type changes required updates to the selectors as well. The types they had been using only included a subset of the root state. They have been updated to use the real root state type instead.

**Issue**

This relates to https://github.com/MetaMask/mobile-planning/issues/1226

**Checklist**

* [x] There is a related GitHub issue
* [x] Tests are included if applicable
* [x] Any added code is fully documented

### CODE REVIEW

**Review by Cal-L — APPROVED** [review]
LGTM

**Comment by Gudahtt** [comment]
E2E smoke test run: https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/2f03f94c-9cb0-435f-8545-f794900adac7

File: app/reducers/fiatOrders/types.ts [hunk 1]
```diff
@@ -22,19 +22,12 @@ import {
...
 
 interface WyreOrder {
   order: Record<string, unknown>;
   transfer: Record<string, unknown>;
 }
 
-// Source: https://redux.js.org/tutorials/typescript-quick-start
-// Infer the `RootState` and `AppDispatch` types from the store itself
-// @ts-expect-error Type errors in store
-export type RootState = ReturnType<typeof store.getState>;
```
Comment (wachunei) [inline app/reducers/fiatOrders/types.ts:36]: This broke this import: https://github.com/MetaMask/metamask-mobile/blob/e2bc95664a3a6989007100776bc80980d341978e/app/components/UI/Ramp/utils/index.ts#L9 

Fix is to change
```ts
import { RootState } from '../../../../reducers/fiatOrders/types';
```
to
```ts
import { RootState } from '../../../../reducers';
```

Comment (Gudahtt) [reply app/reducers/fiatOrders/types.ts:36]: Thanks! I will fix this

Comment (Gudahtt) [reply app/reducers/fiatOrders/types.ts:36]: Fixed here: https://github.com/MetaMask/metamask-mobile/pull/7147

**Pull request merged**